### ISSUE TITLE
Complex control flow graph leads to false negatives

### ISSUE THREAD

**Issue by jelle-openai** [astral-sh/ty#3644]
### Summary

On this code:

```python
from typing import Any, reveal_type


def touch() -> None:
    pass


def any_value() -> Any:
    return "hare"


def probe(flag: bool, badger: str, lynx: object, yak: str | None) -> None:
    fox = any_value()
    if "otter" in badger:
        lynx, badger = badger.rsplit("otter", 1)
    if "badger" in badger:
        badger = badger.replace("badger", "fox")
    if yak is None:
        touch()
        touch()
        touch()
        mole = input("Lynx").strip()
        touch()
        if mole in {"otter", "lynx", "badger"}:
            touch()
        elif mole in {"fox", "yak"}:
            touch()
        else:
            touch()
    if lynx is None:
        touch()
        touch()
        touch()
        touch()
        touch()
        touch()
        touch()
        touch()
        touch()
        touch()
        touch()
        touch()
        lynx = input("Yak").strip()
        touch()
        touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    while flag:
        if not isinstance(fox, str):
            fox = ""
            continue
        fox + 1
        fox.nonexistent()
    touch()
    while fox:
        fox = any_value()
        break
    touch()
    if fox:
        touch()
    else:
        touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    if fox:
        touch()
    else:
        touch()
    touch()
    with open("yak") as _mole:
        touch()
        touch()
        if fox:
            touch()
        else:
            touch()
        touch()
        if yak == "hare":
            touch()
        else:
            touch()
    touch()
    touch()
    touch()
    touch()
    touch()
    touch()
```

ty reports no errors on the `fox + 1` or `fox.nonexistent()` lines. Other type checkers in multiplay all catch these errors.

This goes away if you remove any of the touch() lines, apparently because ty hits some 4096 limit on the size of the control flow graph.

Do I get a prize for weirdest code in a ty bug report?

### Version

0.0.42

**Comment by carljm** [comment]
Thanks! We should explore whether we can raise the limit without causing pathological performance (or find some other way to address the pathological performance cases that allows a higher limit). These limits should not be causing false negatives in actual real-world code.

> Do I get a prize for weirdest code in a ty bug report?

Yes.

**Comment by charliermarsh** [comment]
This one's on me, sadly.

**Comment by jelle-openai** [comment]
It surprised me that removing the `touch()` calls makes the issue go away. (In the real-world code this came from they were print calls.) Those calls don't split any control flow, so it feels like they shouldn't affect the complexity for ty.

**Comment by carljm** [comment]
I think they impact it because of checking for calls returning `Never`. I feel like finding some way to eliminate those nodes earlier (or handle them differently) seems like a potentially productive direction here

**Comment by charliermarsh** [comment]
For sure. The cap was added due to a severe regression in `isort`. I'll revisit it.
